### ISSUE TITLE
XHTTP/H2: per-stream receive window left at the 64 KiB default caps one proxied connection at ~3.6 Mbit/s on a 113 ms path

### ISSUE THREAD

**Issue by f1uddy** [aimalygin/xray-rust#28]
### Summary

On XHTTP over HTTP/2 the per-stream receive window is left at the h2 default (65535 bytes), while the connection window is raised to 16 MiB. Because one proxied TCP connection maps to one XHTTP stream, that default bounds each user connection directly. On a 113 ms path this caps a single connection at ~3.6 Mbit/s.

### Measurement

v0.6.0 embedded through the FFI on an iPhone 14 Pro, core owns the tun fd, VLESS/XHTTP outbound (`mode: packet-up`, `xmux.maxConnections 2-3`) to a CDN origin. RTT to the origin 113-115 ms, measured on the device.

| concurrent proxied connections | aggregate | per connection |
|---|---|---|
| 8 | 28.5 Mbit/s | 3.56 Mbit/s |
| 32 | 119.3 Mbit/s | 3.73 Mbit/s |

Aggregate scales linearly with connection count and the per-second series is flat — no congestion or loss signature. A single TCP flow through the tunnel never exceeds ~3.6 Mbit/s regardless of how idle the path is. 65535 bytes / 0.113 s ≈ 4.6 Mbit/s, which matches once framing and the ack cadence are taken into account.

The same route, same node, same device against Xray-core's Go client gives 140-145 Mbit/s on a single connection; `x/net/http2` defaults there are 4 MiB per stream and 1 GiB per connection.

### Source

`crates/xray-transport/src/stream/xhttp/h2.rs`:

```rust
/// Keeps one slow response from consuming the receive window shared by every
/// XHTTP stream on the connection. Per-stream flow control remains at the
/// HTTP/2 default, so an individual unread body is still tightly bounded.
const CONNECTION_WINDOW_SIZE: u32 = 16 * 1024 * 1024;
```

`connect_h2_with_keepalive` sets `initial_connection_window_size(CONNECTION_WINDOW_SIZE)` and `initial_max_send_streams(1)`; `initial_stream_window_size` is never set, in this file or elsewhere in the crate.

### Question

Is the per-stream default deliberate as a memory bound, and would you consider either making it configurable or deriving it from the observed BDP? A fixed 64 KiB is well matched to a LAN benchmark and to holding many idle flows cheaply, but on a mobile path with 100 ms of latency it becomes the binding constraint on every single connection, which is not visible in the aggregate-throughput benchmarks.

Happy to re-run the measurement against a patched build if that would help.

**Comment by aimalygin** [comment]
Hey, 
Thank you for the detailed issue. I'm working on it. I'll run some tests before fixing it.

**Comment by aimalygin** [comment]
> ### Summary
> On XHTTP over HTTP/2 the per-stream receive window is left at the h2 default (65535 bytes), while the connection window is raised to 16 MiB. Because one proxied TCP connection maps to one XHTTP stream, that default bounds each user connection directly. On a 113 ms path this caps a single connection at ~3.6 Mbit/s.
> 
> ### Measurement
> v0.6.0 embedded through the FFI on an iPhone 14 Pro, core owns the tun fd, VLESS/XHTTP outbound (`mode: packet-up`, `xmux.maxConnections 2-3`) to a CDN origin. RTT to the origin 113-115 ms, measured on the device.
> 
> concurrent proxied connections	aggregate	per connection
> 8	28.5 Mbit/s	3.56 Mbit/s
> 32	119.3 Mbit/s	3.73 Mbit/s
> Aggregate scales linearly with connection count and the per-second series is flat — no congestion or loss signature. A single TCP flow through the tunnel never exceeds ~3.6 Mbit/s regardless of how idle the path is. 65535 bytes / 0.113 s ≈ 4.6 Mbit/s, which matches once framing and the ack cadence are taken into account.
> 
> The same route, same node, same device against Xray-core's Go client gives 140-145 Mbit/s on a single connection; `x/net/http2` defaults there are 4 MiB per stream and 1 GiB per connection.
> 
> ### Source
> `crates/xray-transport/src/stream/xhttp/h2.rs`:
> 
> /// Keeps one slow response from consuming the receive window shared by every
> /// XHTTP stream on the connection. Per-stream flow control remains at the
> /// HTTP/2 default, so an individual unread body is still tightly bounded.
> const CONNECTION_WINDOW_SIZE: u32 = 16 * 1024 * 1024;
> `connect_h2_with_keepalive` sets `initial_connection_window_size(CONNECTION_WINDOW_SIZE)` and `initial_max_send_streams(1)`; `initial_stream_window_size` is never set, in this file or elsewhere in the crate.
> 
> ### Question
> Is the per-stream default deliberate as a memory bound, and would you consider either making it configurable or deriving it from the observed BDP? A fixed 64 KiB is well matched to a LAN benchmark and to holding many idle flows cheaply, but on a mobile path with 100 ms of latency it becomes the binding constraint on every single connection, which is not visible in the aggregate-throughput benchmarks.
> 
> Happy to re-run the measurement against a patched build if that would help.

Could you try this rc https://github.com/aimalygin/xray-rust/releases/tag/v0.6.1-rc.1
